# Проект по SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных:**

**Таблица books**

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

**Задания:**

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [2]:
def showcase(df):
    table = pd.io.sql.read_sql(sql=text(df), con = con)
    return table

In [3]:
books = showcase('''
SELECT * 
FROM books 
''')

display(books.head())

authors = showcase('''
SELECT * 
FROM authors 
''')

display(authors.head())

publishers = showcase('''
SELECT * 
FROM publishers 
''')

display(publishers.head())

ratings = showcase('''
SELECT * 
FROM ratings 
''')

display(ratings.head())

reviews = showcase('''
SELECT * 
FROM reviews 
''')

display(reviews.head())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**Вывод:** базы данных подключены

## Подсчёт книг вышедших после 1 января 2000 года;

In [4]:
books_count = '''
            SELECT COUNT(book_id) as books_count
            FROM books
            WHERE publication_date > '2000-01-01'
            '''

showcase(books_count)

books_count
0          819

**Вывод:** после 1 января 2000 года вышло 819 книг.

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Правильно!
    
Отмечу, что условие фильтрации `publication_date >= '2000-01-01' ` также было бы правильным, хоть и привело бы к другому результату - в реальных проектах о том, включается ли дата в диапазон или нет, всегда стоит уточнять заранее:)

</div>

##  Количество обзоров и средняя оценка каждой книги

In [5]:
books_reviews = '''
WITH book_stats AS (
    SELECT 
           b.book_id,
           b.title,
           COUNT(DISTINCT rev.review_id) AS count_reviews,
           AVG(rat.rating) AS avg_rating
    FROM books b
    LEFT JOIN ratings rat ON b.book_id = rat.book_id
    LEFT JOIN reviews rev ON b.book_id = rev.book_id
    GROUP BY b.book_id
)

SELECT * 
FROM (
    SELECT * 
    FROM book_stats
    ORDER BY avg_rating DESC, count_reviews DESC
    LIMIT 50
) AS top_by_rating

UNION ALL

SELECT * 
FROM (
    SELECT * 
    FROM book_stats
    ORDER BY count_reviews DESC, avg_rating DESC
    LIMIT 50
) AS top_by_reviews
'''


df = showcase(books_reviews)


top_by_rating = df.iloc[:50]
top_by_reviews = df.iloc[50:].reset_index(drop=True)

display(top_by_rating)
display(top_by_reviews)

book_id                                              title  count_reviews  \
0        17                      A Dirty Job (Grim Reaper  #1)              4   
1       444       Moneyball: The Art of Winning an Unfair Game              3   
2       553            School's Out—Forever (Maximum Ride  #2)              3   
3       136  Captivating: Unveiling the Mystery of a Woman'...              2   
4        62                                      Alas  Babylon              2   
5       938                          Triptych (Will Trent  #1)              2   
6       347  In the Hand of the Goddess (Song of the Liones...              2   
7       598               Stolen (Women of the Otherworld  #2)              2   
8       229           Evil Under the Sun (Hercule Poirot  #24)              2   
9       972  Wherever You Go  There You Are: Mindfulness Me...              2   
10      902  The War of Art: Break Through the Blocks & Win...              2   
11       76                                        Angels Fall              2   
12      913                                 The Woman in Black              2   
13      516                       Plum Island (John Corey  #1)              2   
14      642                 The Big Bad Wolf  (Alex Cross  #9)              2   
15      297                                         Hard Times              2   
16      421                                        Marvel 1602              2   
17      699  The Demon-Haunted World: Science as a Candle i...              2   
18      901  The Walking Dead  Book One (The Walking Dead #...              2   
19      993                                              Women              2   
20      967               Welcome to Temptation (Dempseys  #1)              2   
21       55          A Woman of Substance (Emma Harte Saga #1)              2   
22      459                               My Name Is Asher Lev              2   
23      610                           Tai-Pan (Asian Saga  #2)              2   
24      213                      Emily of New Moon (Emily  #1)              2   
25      441                  Misty of Chincoteague (Misty  #1)              2   
26      469                           Neil Gaiman's Neverwhere              2   
27      518               Pop Goes the Weasel (Alex Cross  #5)              2   
28      169  Crucial Conversations: Tools for Talking When ...              2   
29      513               Piercing the Darkness (Darkness  #2)              2   
30      418                                              March              2   
31      182                                         Dead Souls              2   
32      390                                    Light in August              2   
33      275                                          Geek Love              2   
34      321                                Homage to Catalonia              2   
35      732  The Ghost Map: The Story of London's Most Terr...              2   
36       86      Arrows of the Queen (Heralds of Valdemar  #1)              2   
37       57                    Act of Treason (Mitch Rapp  #9)              2   
38       20              A Fistful of Charms (The Hollows  #4)              2   
39      224                                      Evening Class              2   
40      625  The Adventures of Tom Sawyer and Adventures of...              1   
41      330  How to Be a Domestic Goddess: Baking and the A...              1   
42      599  Stone of Farewell (Memory  Sorrow  and Thorn  #2)              1   
43      672   The Cat in the Hat and Other Dr. Seuss Favorites              0   
44      195                 Dragonfly in Amber (Outlander  #2)              4   
45      524             Preludes & Nocturnes (The Sandman  #1)              3   
46      327                                    House of Leaves              4   
47      912  The Witching Hour (Lives of the Mayfair Witche...              3   
48      411      Lover Revealed (Bl

book_id                                              title  count_reviews  \
0       948                            Twilight (Twilight  #1)              7   
1       302  Harry Potter and the Prisoner of Azkaban (Harr...              6   
2       299  Harry Potter and the Chamber of Secrets (Harry...              6   
3       656                                     The Book Thief              6   
4       734                                   The Glass Castle              6   
5       750                The Hobbit  or There and Back Again              6   
6       497                          Outlander (Outlander  #1)              6   
7       695  The Curious Incident of the Dog in the Night-Time              6   
8       779  The Lightning Thief (Percy Jackson and the Oly...              6   
9       963                                Water for Elephants              6   
10      696             The Da Vinci Code (Robert Langdon  #2)              6   
11      673                             The Catcher in the Rye              6   
12      627                                      The Alchemist              6   
13      854                                           The Road              6   
14      733                          The Giver (The Giver  #1)              6   
15      207                                    Eat  Pray  Love              6   
16      581                     Snow Flower and the Secret Fan              5   
17       74                           And Then There Were None              5   
18      884                                The Thirteenth Tale              5   
19      284                                 Gone with the Wind              5   
20      722  The Fellowship of the Ring (The Lord of the Ri...              5   
21       15     A Clash of Kings  (A Song of Ice and Fire  #2)              5   
22      830                               The Poisonwood Bible              5   
23      862  The Sea of Monsters (Percy Jackson and the Oly...              5   
24       51                           A Tree Grows in Brooklyn              5   
25      300  Harry Potter and the Half-Blood Prince (Harry ...              5   
26      691                          The Count of Monte Cristo              5   
27      304                         Hatchet (Brian's Saga  #1)              5   
28      426                                Memoirs of a Geisha              5   
29      399                                       Little Women              5   
30      301  Harry Potter and the Order of the Phoenix (Har...              5   
31      869                                        The Shining              5   
32      837                                 The Princess Bride              5   
33      461                                 My Sister's Keeper              5   
34      162                                           Coraline              5   
35      374                                 Kafka on the Shore              5   
36      364                                          Jane Eyre              5   
37      408                        Love in the Time of Cholera              5   
38      472                      Neverwhere (London Below  #1)              5   
39      950                                Uglies (Uglies  #1)              5   
40      577                                Slaughterhouse-Five              5   
41      417                           Man's Search for Meaning              5   
42      824                         The Picture of Dorian Gray              5   
43       84    Anne of Green Gables (Anne of Green Gables  #1)              5   
44       70                  American Gods (American Gods  #1)              5   
45      358                                      Into the Wild              5   
46      124                                    Brave New World              5   
47      478                                   Nineteen Minutes              5   
48      320                        

**Вывод:** Максимум обзоров Twilight - 7, однако рейдинг 3.66. Среди книг с высоким рейтингом (5) максималньое количество отзывов - A Dirty Job (Grim Reaper #1) - 4. Две книги с высоким рейтингом имеют 3 обзора. У остальных книг с высоким рейтингом от 0 до 2 обзоров. Среди книг у которых более 4х обзоров максимальный рейтинг у книги Snow Flower and the Secret Fan -  4.636364.

Можем сделать вывод, что чем больше обзоров у книги, тем сложнее ей получить максимальную оценку, потому что одна негативная оценка может сильно опустить средний рейтинг. Вероятно обзоры влияют на восприятие и оценку обычных пользователей. 

## Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [6]:
publisher_50 = '''
               SELECT pub.publisher AS publisher,
                      COUNT(b.book_id) AS books_cnt
               FROM publishers pub
               LEFT JOIN books b ON pub.publisher_id = b.publisher_id
               WHERE b.num_pages > 50
               GROUP BY publisher
               ORDER BY books_cnt DESC
               LIMIT 5
               '''

showcase(publisher_50)

publisher  books_cnt
0             Penguin Books         42
1                   Vintage         31
2  Grand Central Publishing         25
3          Penguin Classics         24
4                    Bantam         19

**Вывод:** больше всего книг больше 50 страниц выпустило издательство Penguin Books - 42 книги.

## Автор с самой высокой средней оценкой книг — учитываются только книги с 50 и более оценками

In [7]:
author_rating = '''
WITH book_ratings AS (
     SELECT 
         b.book_id,
         b.author_id,
         AVG(r.rating) AS avg_rating,
         COUNT(r.rating_id) AS rating_count
     FROM books b
     JOIN ratings r ON b.book_id = r.book_id
     GROUP BY b.book_id, b.author_id
     HAVING COUNT(r.rating_id) >= 50
),
author_avg_rating AS (
    SELECT 
        author_id,
        AVG(avg_rating) AS avg_author_rating
    FROM book_ratings
    GROUP BY author_id
)
SELECT 
    a.author,
    ar.avg_author_rating
FROM author_avg_rating ar
JOIN authors a ON ar.author_id = a.author_id
ORDER BY ar.avg_author_rating DESC
LIMIT 3;
'''

showcase(author_rating)

author  avg_author_rating
0         J.K. Rowling/Mary GrandPré           4.283844
1  Markus Zusak/Cao Xuân Việt Khương           4.264151
2                     J.R.R. Tolkien           4.258446

**Вывод:** Автор с самым выскоим средним рейтингом: J.K. Rowling/Mary GrandPré. Средний рейтинг: 4.28

## Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [8]:
avg_reviews = '''
WITH user_reviews AS (
     SELECT r.username,
            COUNT(r.review_id) AS review_count
     FROM reviews r
     JOIN(SELECT username
          FROM ratings
          GROUP BY username
          HAVING COUNT(rating_id)>48
     ) u ON r.username = u.username
     GROUP BY r.username
     )
SELECT AVG(review_count) AS avg_reviews
FROM user_reviews;
'''

showcase(avg_reviews)

avg_reviews
0         24.0

**Вывод:** среднее количество обзоров среди пользователей, поставивших более 48 оценок - 24. 

## Общие выводы:

- После 1 января 2000 года вышло 819 книг.
- Максимум обзоров Twilight - 7, однако рейдинг 3.66. Среди книг с высоким рейтингом (5) максималньое количество отзывов - A Dirty Job (Grim Reaper #1) - 4. Две книги с высоким рейтингом имеют 3 обзора. У остальных книг с высоким рейтингом от 0 до 2 обзоров. Среди книг у которых более 4х обзоров максимальный рейтинг у книги Snow Flower and the Secret Fan - 4.636364. Можем сделать вывод, что чем больше обзоров у книги, тем сложнее ей получить максимальную оценку, потому что одна негативная оценка может сильно опустить средний рейтинг. Вероятно обзоры влияют на восприятие и оценку обычных пользователей.
- Больше всего книг больше 50 страниц выпустило издательство Penguin Books - 42 книги.
- Автор с самым выскоим средним рейтингом: J.K. Rowling/Mary GrandPré. Средний рейтинг: 4.28
- Среднее количество обзоров среди пользователей, поставивших более 48 оценок - 24. 